### PR TITLE
fix: utils leaking Scene state (#6461

### PR DESCRIPTION
**Pull request by dwelle**
fix https://github.com/excalidraw/excalidraw/issues/6453

In https://github.com/excalidraw/excalidraw/pull/6443 we've introduced a hacky solution for export utils not having access to elements when retrieving them through our Scene class. But, it broke some logic because the ad-hoc Scene polluted references with elements that weren't actually on the scene and shouldn't be considered (the `id:element` mapping would replace the actual elements on canvas with the mock elements).

This PR fixes it by duplicating the elements (and most importantly regenerating their ids) before populating the ad-hoc Scene in export utils.

The newly introduced `duplicateElements` helper ensures that all ids including that of bindings are regenerated but kept consistent (so that bindings are preserved). We'll actually use this helper later (I believe there's one unrelated bug due to incorrect duplication).

---

### Summary
🎨🧪♻️

Added a new feature to duplicate multiple elements and improved the export functions. The main changes are:

> _We'll `exportToSvg` and `exportToBlob` with ease_
> _We'll clone and duplicate our elements as we please_
> _We'll test our code with jest and jsdom, one, two, three_
> _We're the jolly coders of the `utils` library_

### Walkthrough
*  Rename and refactor `deepCopyElement` function to ensure type safety and immutability, and add a wrapper function to perform type check ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-205831ba144cafaba7bdcaab8e1affeb0eb97a4da50714294f73d307347e5d8eL360-R382), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-205831ba144cafaba7bdcaab8e1affeb0eb97a4da50714294f73d307347e5d8eL381-R394), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-205831ba144cafaba7bdcaab8e1affeb0eb97a4da50714294f73d307347e5d8eL391-R439))
* Add `duplicateElements` function to clone an array of elements, regenerating their ids and group ids, and updating their bindings and containers ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-205831ba144cafaba7bdcaab8e1affeb0eb97a4da50714294f73d307347e5d8eL413-R459))
* Update import statements to include `duplicateElements` and `arrayToMap` functions ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-eb3c7d3b2122429a45c47066dce3aea85ea40e385f8f7ce53e32484c78abc311L1-R6), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-205831ba144cafaba7bdcaab8e1affeb0eb97a4da50714294f73d307347e5d8eL16-R21), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L18-R35))
* Modify `exportToCanvas`, `exportToBlob`, and `exportToSvg` functions to use `passElementsSafely` helper function instead of creating a new `Scene` instance and replacing all elements, to handle cases where elements are not cached by the `Scene` instance ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L47-R66), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L125-R136), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L181-R202))
* Modify `exportToBlob` and `exportToClipboard` functions to pass the original, uncloned elements to the `serializeAsJSON` and `copyToClipboard` functions, to ensure that the ids of the elements are stable and consistent when embedding the scene data into the blob and clipboard metadata ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L153-R157), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L206-L213), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-38000639f2f5947b0254239a2d498f800cfba3ce0ccdd9a10eec865e5ca99fa9L228-R227))
* Modify `exportToSvg` function in `src/scene/export.ts` to accept an optional `opts` argument, to allow passing a custom `serializeAsJSON` function ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-91b6e2243746c6661270480d9d05856a54d6af764eabe54f1f2df33f082b7b47R93-R95), [link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-91b6e2243746c6661270480d9d05856a54d6af764eabe54f1f2df33f082b7b47L106-R111))
* Add `describe` block to group the existing tests for duplicating single elements under a common label, and add new tests for duplicating multiple elements ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-eb3c7d3b2122429a45c47066dce3aea85ea40e385f8f7ce53e32484c78abc311L18-R324))
* Add a new test file `src/tests/packages/utils.unmocked.test.ts` to test the embedding of scene data into the exported svg and blob formats, using the `decodeSvgMetadata` and `decodePngMetadata` functions, the custom `serializeAsJSON` function, and the `API` helper ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-1a88936accdafb66d57c0291966be78140226f3d8b4d92baa0e245d87acf888dR1-R63))
* Remove the global function `window.wrapSelected` before merging, as it is not part of the pull request and was added for debugging purposes ([link](https://github.com/excalidraw/excalidraw/pull/6461/files?diff=unified&w=0#diff-205831ba144cafaba7bdcaab8e1affeb0eb97a4da50714294f73d307347e5d8eR492-R600))

### CODE REVIEW

File: src/tests/packages/utils.unmocked.test.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { decodePngMetadata, decodeSvgMetadata } from "../../data/image";
```
Comment (ad1992) [inline src/tests/packages/utils.unmocked.test.ts:1]: lets rename this file to `utils.test.ts` since it has the tests for the same file

File: src/tests/packages/utils.unmocked.test.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import { ImportedDataState } from "../../data/types";
+import * as utils from "../../packages/utils";
+import { API } from "../helpers/api";
+
+describe("embedding scene data", () => {
+  describe("exportToSvg", () => {
```
Comment (ad1992) [inline src/tests/packages/utils.unmocked.test.ts:7]: Do we need a separate describe block for this ? Instead the it block could be renamed to `shouldn't modify the elements when exporting to svg`

File: src/element/newElement.ts [hunk 3]
```diff
@@ -449,3 +489,112 @@ export const duplicateElement = <TElement extends Mutable<ExcalidrawElement>>(
...
+    }
+
+    clonedElements.push(clonedElement);
+  }
+
+  return clonedElements;
+};
+
+// @ts-ignore
+window.wrapSelected = (...args) => {
```
Comment (ad1992) [inline src/element/newElement.ts]: what is this for ?

File: src/element/newElement.test.ts [hunk 4]
```diff
@@ -15,79 +16,310 @@ const assertCloneObjects = (source: any, clone: any) => {
...
-    strokeColor: "#000000",
-    backgroundColor: "transparent",
-    fillStyle: "hachure",
-    strokeWidth: 1,
-    strokeStyle: "solid",
-    roundness: { type: ROUNDNESS.PROPORTIONAL_RADIUS },
-    roughness: 1,
-    opacity: 100,
-  });
+describe("duplicating single elements", () => {
```
Comment (ad1992) [inline src/element/newElement.test.ts:19]:
```suggestion
describe("duplicate single elements", () => {
```

File: src/element/newElement.test.ts [hunk 5]
```diff
@@ -15,79 +16,310 @@ const assertCloneObjects = (source: any, clone: any) => {
...
+      fillStyle: "hachure",
+      strokeWidth: 1,
+      strokeStyle: "solid",
+      roundness: { type: ROUNDNESS.PROPORTIONAL_RADIUS },
+      roughness: 1,
+      opacity: 100,
+    });
+
+    // @ts-ignore
+    element.__proto__ = { hello: "world" };
```
Comment (ad1992) [inline src/element/newElement.test.ts:36]: lets add a comment about this

File: src/element/newElement.test.ts [hunk 6]
```diff
@@ -15,79 +16,310 @@ const assertCloneObjects = (source: any, clone: any) => {
...
-    roundness: null,
-    roughness: 1,
-    opacity: 100,
-    text: "hello",
-    fontSize: 20,
-    fontFamily: FONT_FAMILY.Virgil,
-    textAlign: "left",
-    verticalAlign: "top",
+describe("duplicating multiple elements", () => {
+  it("duplicateElements should clone bindings", () => {
```
Comment (ad1992) [inline src/element/newElement.test.ts:99]:
```suggestion
  it("should clone bindings", () => {
```
As the describe block conveys the block is about duplicating elements

File: src/element/newElement.test.ts [hunk 7]
```diff
@@ -15,79 +16,310 @@ const assertCloneObjects = (source: any, clone: any) => {
...
-    strokeStyle: "solid",
-    roundness: null,
-    roughness: 1,
-    opacity: 100,
-    text: "hello",
-    fontSize: 20,
-    fontFamily: FONT_FAMILY.Virgil,
-    textAlign: "left",
-    verticalAlign: "top",
+describe("duplicating multiple elements", () => {
```
Comment (ad1992) [inline src/element/newElement.test.ts:98]:
```suggestion
describe("Test duplicateElements", () => {
```

File: src/element/newElement.ts [hunk 8]
```diff
@@ -388,14 +401,41 @@ export const deepCopyElement = (val: any, depth: number = 0) => {
...
+      objectType !== "[object Object]" &&
+      objectType !== "[object Array]" &&
+      objectType.startsWith("[object ")
+    ) {
+      console.warn(
+        `_deepCloneElement: unexpected object type ${objectType}. This value will not be cloned!`,
+      );
+    }
+  }
+
```
Comment (ad1992) [inline src/element/newElement.ts:420]: Why is this needed ? can you add a comment ?

File: src/element/newElement.test.ts (hunk 5 above)
Comment (dwelle) [reply src/element/newElement.test.ts:36]: not my test :) and I have no clue why we're testing for it

File: src/element/newElement.ts (hunk 3 above)
Comment (dwelle) [reply src/element/newElement.ts]: removed

File: src/tests/packages/utils.unmocked.test.ts (hunk 1 above)
Comment (dwelle) [reply src/tests/packages/utils.unmocked.test.ts:1]: can't, already exists. I have spent some time yesterday figuring out how to have a test file that both tests with mocked and unmocked modules, and didn't figure it out.

File: src/tests/packages/utils.unmocked.test.ts (hunk 2 above)
Comment (dwelle) [reply src/tests/packages/utils.unmocked.test.ts:7]: doesn't matter

File: src/element/newElement.ts (hunk 8 above)
Comment (dwelle) [reply src/element/newElement.ts:420]: it's explained in the function's tsdoc, but I'll put a comment here as well

File: src/tests/packages/utils.unmocked.test.ts (hunk 1 above)
Comment (ad1992) [reply src/tests/packages/utils.unmocked.test.ts:1]: In that case, lets add a comment on top of this file

File: src/tests/packages/utils.unmocked.test.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    it("embedding scene data shouldn't modify them", async () => {
```
Comment (ad1992) [inline src/tests/packages/utils.unmocked.test.ts:8]:
```suggestion
    it("shouldn't modify the scene", async () => {
```
If you want to keep the describe block, lets rename this to be more clear and avoid redundancy

File: src/tests/packages/utils.unmocked.test.ts [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      expect(sourceElements.map((x) => x.id)).toEqual(
+        importedData.elements?.map((el) => el.id),
+      );
+    });
+  });
+
+  // skipped because we can't test png encoding right now
+  // (canvas.toBlob not supported in jsdom)
+  describe.skip("exportToBlob", () => {
+    it("embedding scene data shouldn't modify them", async () => {
```
Comment (ad1992) [inline src/tests/packages/utils.unmocked.test.ts:38]:
```suggestion
    it("shouldn't modify the scene", async () => {
```

File: src/tests/packages/utils.unmocked.test.ts [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import { decodePngMetadata, decodeSvgMetadata } from "../../data/image";
+import { ImportedDataState } from "../../data/types";
+import * as utils from "../../packages/utils";
+import { API } from "../helpers/api";
+
+describe("embedding scene data", () => {
```
Comment (ad1992) [inline src/tests/packages/utils.unmocked.test.ts:6]:
```suggestion
describe("Test embedding scene data", () => {
```
The block renaming suggestions are mainly for consistency and more descriptive tests - I will add some linters for the test so we can try to follow the similar pattern across all tests.

File: src/tests/packages/utils.unmocked.test.ts (hunk 2 above)
Comment (ad1992) [inline src/tests/packages/utils.unmocked.test.ts:7]:
```suggestion
  describe("When using exportToSvg", () => {
```

File: src/element/newElement.test.ts (hunk 5 above)
Comment (dwelle) [reply src/element/newElement.test.ts:36]: btw, you should be reviewing with whitespace disabled so it's clear what's actually a change, and what's whitespace noop

**Comment by dwelle** [comment]
@ad1992 Thanks for the review 🙏

The test naming discussed on chat, and we'll resolve later.
